napari/layers/volume/volume.py

```diff
@@ -0,0 +1,224 @@
+            else:
+                self._data_view = np.zeros((1, 1, 1))
+            self._data_thumbnail = self._data_view
+
+            # Set colormap
+            self._colormap_name = ''
+            self.colormap = colormap
+
+            # Set update flags
+            self._need_display_update = False
+            self._need_visual_update = False
+
+            # Re intitialize indices
+            self._indices = (
+                slice(None, None, None),
+                slice(None, None, None),
+                slice(None, None, None),
+            )
+
+            # Trigger generation of view slice and thumbnail
+            self._set_view_slice()
+
+    @property
+    def data(self):
+        """array: Image data."""
+        return self._data
+
+    @data.setter
+    def data(self, data):
+        self._data = data
+        if self.multichannel:
+            self._multichannel = is_multichannel(data.shape)
+        self.events.data()
+        self.refresh()
+
+    def _get_shape(self):
+        if self.multichannel:
+            return self.data.shape[:-1]
+        return self.data.shape
+
+    @property
+    def multichannel(self):
+        """bool: Whether the image is multichannel."""
+        return self._multichannel
+
+    @multichannel.setter
+    def multichannel(self, multichannel):
+        if multichannel is False:
+            self._multichannel = multichannel
+        else:
+            # If multichannel is True or None then guess if multichannel
+            # allowed or not, and if allowed set it to be True
+            self._multichannel = is_multichannel(self.data.shape)
+        self.refresh()
+
+    @property
+    def colormap(self):
+        """2-tuple of str, vispy.color.Colormap: colormap for luminance images.
+        """
+        return self._colormap_name, self._node.cmap
+
+    @colormap.setter
+    def colormap(self, colormap):
+        name = '[unnamed colormap]'
+        if isinstance(colormap, str):
+            name = colormap
+        elif isinstance(colormap, tuple):
+            name, cmap = colormap
+            self._colormaps[name] = cmap
+        elif isinstance(colormap, dict):
+            self._colormaps.update(colormap)
+            name = list(colormap)[0]  # first key in dict
+        elif isinstance(colormap, vispy.color.Colormap):
+            name = increment_unnamed_colormap(
+                name, list(self._colormaps.keys())
+            )
+            self._colormaps[name] = colormap
+        else:
+            warnings.warn(f'invalid value for colormap: {colormap}')
+            name = self._colormap_name
+        self._colormap_name = name
+        self._node.cmap = self._colormaps[name]
+        self.events.colormap()
+
+    @property
+    def colormaps(self):
+        """tuple of str: names of available colormaps."""
+        return tuple(self._colormaps.keys())
+
+    def _set_view_slice(self):
+        """Set the view given the indices to slice with."""
+        indices = (
```

Comment:
this shouldn't be needed anymore, you should be able to use `self.indices` after having done `dims.set_display` in my comment above

Reply:
these need to be dropped

Reply:
please make this match the image layer, but with the right number of dims
```python
        indices = list(self.indices)
        indices[:-3] = np.clip(
            indices[:-3], 0, np.subtract(self.shape[:-3], 1)
        )
```
this should make your `4D` test example pass